MM Build 2025 - Skadoosh - Serve plugin metrics to standard /metrics endpoint

cpoile opened the pull request:
#### Summary
- Part of the MM Build 2025 event, serve plugin metrics to standard /metrics endpoint
- Namespace them based on the plugin name
- Add tests

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-68436


#### Release Note
```release-note
Collecting plugin metrics, namespacing them, and serving on the standarad /metrics endpoint
```




## Change Impact: 🟡 Medium

**Regression Risk:** The change modifies the central /metrics endpoint to aggregate and rewrite Prometheus text from plugins (stripping Accept-Encoding, injecting/overwriting plugin_id, and re-encoding), which can affect scrape behavior and response correctness; comprehensive tests were added for parsing, feature-flag-off, non-200 plugin responses, and empty bodies, but edge cases like Accept-Encoding/gzip handling remain less directly covered.  
**QA Recommendation:** Moderate manual QA: validate Prometheus scrapes with PluginMetricsCollection on/off, confirm plugin metrics are present and namespaced (including replacing existing plugin_id), verify malformed plugin output is skipped without panic, and exercise gzip/Accept-Encoding paths.

*Generated by CodeRabbitAI*

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Change Impact: 🟡 Medium

**Regression Risk:** Changes modify the central /metrics aggregation path (stripping Accept-Encoding, capturing/re-encoding Prometheus text, and injecting/overwriting plugin_id labels), creating plausible edge cases for Prometheus scrapes and HTTP encoding behavior; the feature flag defaults to OFF which limits immediate exposure, and tests exercise many paths though gzip/Accept-Encoding behavior has lighter coverage.  
**QA Recommendation:** Perform manual QA: (1) validate scrapes with AggregatePluginMetrics toggled on/off, (2) confirm plugin_id injection/replacement for single and multiple plugins, (3) verify malformed or non-200 plugin outputs are skipped without panics, and (4) exercise gzip/Accept-Encoding paths to ensure no regressions.

*Generated by CodeRabbitAI*
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

cpoile committed:
gather plugin metrics and serve on /metrics

cpoile committed:
tests

cpoile committed:
fix recent conflicts from master

cpoile committed:
fix linting

bgardner8008 reviewed (approved):
This looks great to me, nice implementation and really cool integration tests.

I note the test metrics from the plugs don't contain any labels, so hopefully the code to add label to existing labels works. I suppose it could be tested by adding a second plugin metric with a label. But it would be much simpler to just have a unit test for addPluginLabelToMetrics.

lieut-data reviewed (changes requested):
Thanks @cpoile, this will really move the needle on our Grafana efforts! A few thoughts below.

lieut-data committed:
Merge remote-tracking branch 'origin/master' into skadoosh

lieut-data committed:
address review feedback: feature flag, route guard, and unit tests for plugin metrics

- Add PluginMetricsCollection feature flag (default true) to allow disabling plugin metric collection without a deploy
- Only wrap the /metrics handler with plugin metric collection (not arbitrary plugin routes, which would cause double-listing)
- Extract wrapping logic into wrapMetricsHandler for clarity
- Add unit tests for addPluginLabelToMetrics covering no-label, existing-label, comments, empty lines, and multiple metrics cases

lieut-data commented on server/channels/app/platform/metrics.go:
```diff
@@ -241,8 +243,95 @@ func (pm *platformMetrics) servePluginMetricsRequest(w http.ResponseWriter, r *h
 func (ps *PlatformService) HandleMetrics(route string, h http.Handler) {
 	if ps.metrics != nil {
-		ps.metrics.router.Handle(route, h)
+		wrappedHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Added `PluginMetricsCollection` feature flag (default `true`) — the wrapping now only happens when the flag is on.

lieut-data commented on server/channels/app/platform/metrics.go:
```diff
@@ -241,8 +243,95 @@ func (pm *platformMetrics) servePluginMetricsRequest(w http.ResponseWriter, r *h
+			}
+		})
+
+		ps.metrics.router.Handle(route, wrappedHandler)
```
Good catch — fixed. The wrapper is now only applied when `route == "/metrics"`, so any other routes registered via `HandleMetrics` are passed through unchanged.

lieut-data commented on server/channels/app/platform/metrics.go:
```diff
@@ -241,8 +243,95 @@ func (pm *platformMetrics) servePluginMetricsRequest(w http.ResponseWriter, r *h
+		// Find the metric name and value parts
+		// Format: metric_name{labels} value
+		// or: metric_name value
+		if idx := strings.IndexAny(line, "{ "); idx != -1 {
```
Considered it — `prometheus/common` is already an indirect dep, but using `expfmt` to parse, relabel, and re-encode each metric family would add meaningful complexity for what is essentially a straightforward label injection into a well-specified text format. The string manipulation is compact and the new unit tests pin the exact behavior. Happy to switch if you'd prefer the library approach.

lieut-data commented on server/channels/app/platform/metrics_test.go:
```diff
@@ -89,3 +89,8 @@ func TestMetricsRouter(t *testing.T) {
 	}
 }
+
+// NOTE: Integration tests for HandleMetrics with plugin ServeMetrics hooks
```
Added `TestAddPluginLabelToMetrics` covering: metric without labels, metric with existing labels, HELP/TYPE comment preservation, empty line preservation, and multiple metrics.

lieut-data committed:
fix integration tests to use /metrics route so wrapMetricsHandler runs

lieut-data commented on server/channels/app/plugin_hooks_test.go:
```diff
@@ -2076,6 +2075,257 @@ func TestChannelHasBeenCreated(t *testing.T) {
+
+		// Verify both server and plugin metrics are present
+		assert.Contains(t, bodyStr, "server_test_metric 100", "Response should contain server metrics")
+		assert.Contains(t, bodyStr, "plugin_test_metric{plugin_id=\""+pluginID+"\"} 42", "Response should contain plugin metrics with plugin_id label")
```
Fixed — all three subtests now use `/metrics` so `wrapMetricsHandler` runs.

lieut-data committed:
switch addPluginLabelToMetrics to use expfmt for correct label injection

Replace string manipulation with expfmt parse→mutate→re-encode, which
correctly handles all metric types, timestamps, and malformed input.
Output is normalized: blank lines are dropped and TYPE headers are
injected for undeclared metrics. Update unit tests accordingly.

lieut-data committed:
refine addPluginLabelToMetrics: log warnings, drop sort, use assert.Contains

- Log a warning (instead of silently discarding) on parse or encode errors,
  returning empty string in both cases
- Remove name sorting — metric family order is irrelevant to Prometheus scrapers
- Switch unit test assertions to assert.Contains per line so tests are
  order-independent

lieut-data committed:
replace strPtr helper with model.NewPointer

lieut-data reviewed (approved):
Will test locally before merging.

lieut-data committed:
fix gofmt: remove extra blank line

lieut-data committed:
replace existing plugin_id label instead of appending a duplicate

If a plugin already exports a plugin_id label, overwrite it rather than
appending a second one, which would produce invalid Prometheus output.
Add test coverage for the replace path.

lieut-data committed:
add missing test coverage per mattermost-build feedback

- Unit test: malformed input returns empty string without panicking
- Integration: PluginMetricsCollection=false excludes plugin metrics
- Integration: plugin returning non-200 status excluded from response
- Integration: plugin returning empty body excluded from response

lieut-data committed:
rename PluginMetricsCollection to AggregatePluginMetrics, default false

lieut-data committed:
address coderabbit feedback: propagate scrape context and enable flag in negative-path tests

lieut-data committed:
avoid double GetPluginsEnvironment call; skip newline when metrics are empty

amyblais commented:
/cherry-pick release-11.7

mattermost-build commented:
Cherry pick is scheduled.

amyblais commented:
/cherry-pick release-11.7

mattermost-build commented:
Cherry pick is scheduled.

mattermost-build commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
hostfile_replace_entries: mkstemp: Read-only file system
update_known_hosts: hostfile_replace_entries failed for /app/.ssh/known_hosts: Read-only file system
From github.com:mattermost/mattermost
   b8a902866f..87d1ba593e  MM-67755-wysiwyg-editor-v2 -> upstream/MM-67755-wysiwyg-editor-v2
   873298d43e..d4b15753c6  MM-68503-BE-abac-mask-save-path-masking -> upstream/MM-68503-BE-abac-mask-save-path-masking
   50170355cb..f90080f0dd  MM-68506-FE-abac-mask-fe-table-editor-cel-and-e2e -> upstream/MM-68506-FE-abac-mask-fe-table-editor-cel-and-e2e
   1daf0e705e..b2c2721ba5  MM-68649              -> upstream/MM-68649
 + b1bec2343a...43907e6f4c MM-68663.azure.admin  -> upstream/MM-68663.azure.admin  (forced update)
 * [new branch]            MM-68663.azure.admin.clean -> upstream/MM-68663.azure.admin.clean
   b93905d90d..5edaa17f49  MM-68693              -> upstream/MM-68693
   33ef180235..a06e9ecd7d  MM-68779-mbe-phase-8a -> upstream/MM-68779-mbe-phase-8a
   5487c99a1e..490e9a411b  MM-68780-mbe-phase-8b -> upstream/MM-68780-mbe-phase-8b
 * [new branch]            MM-68781-mbe-phase-8d -> upstream/MM-68781-mbe-phase-8d
 * [new branch]            MM-68782-mbe-phase-8e -> upstream/MM-68782-mbe-phase-8e
 * [new branch]            MM-68783-mbe-phase-8f -> upstream/MM-68783-mbe-phase-8f
 * [new branch]            MM-68785-mbe-phase-8h -> upstream/MM-68785-mbe-phase-8h
 * [new branch]            MM-68838_ping_fast_restore_remote -> upstream/MM-68838_ping_fast_restore_remote
   588a95b39c..1bddde7263  automated-cherry-pick-of-mattermost-#36332-upstream-release-10.11 -> upstream/automated-cherry-pick-of-mattermost-#36332-upstream-release-10.11
   9e1a9fc1b7..abc961f4e8  cherry-pick-36442-release-11.6 -> upstream/cherry-pick-36442-release-11.6
 * [new branch]            cherry-pick-shared-channel-plugin-api-11.7 -> upstream/cherry-pick-shared-channel-plugin-api-11.7
 * [new branch]            chore-pin-commit-sha  -> upstream/chore-pin-commit-sha
 * [new branch]            cursor/bcaab566       -> upstream/cursor/bcaab566
   91fbdffdf7..4ebf6657be  cursor/fix-group-role-immutable-mutation-d5b7 -> upstream/cursor/fix-group-role-immutable-mutation-d5b7
 * [new branch]            cursor/fix-msg-slash-wrong-channel-37c0 -> upstream/cursor/fix-msg-slash-wrong-channel-37c0
   517ac4f2fa..63fbcce7e2  cursor/harden-cert-removals-cdb8 -> upstream/cursor/harden-cert-removals-cdb8
 + c1a44151fb...1c52bbb1cf cursor/mm-68763-fdc9  -> upstream/cursor/mm-68763-fdc9  (forced update)
 * [new branch]            cursor/mm-68764-decd  -> upstream/cursor/mm-68764-decd
 * [new branch]            e2e-npm-cache         -> upstream/e2e-npm-cache
 + d43be31724...f05d51d5a1 generate-cypress-role.js -> upstream/generate-cypress-role.js  (forced update)
   3fa5d7fb84..f2dae1de00  manual-cherry-pick-gitlab-1.12.2-on-11.6 -> upstream/manual-cherry-pick-gitlab-1.12.2-on-11.6
   d4fc0ecb1c..8eb97fa6c3  master                -> upstream/master
   bf1f83ed08..c242a118f8  mm-68151-update-server-deps -> upstream/mm-68151-update-server-deps
 * [new branch]            mm-68154-update-imagemeta -> upstream/mm-68154-update-imagemeta
   9d2004dd81..f9a8168e76  release-11.5          -> upstream/release-11.5
 * [new branch]            release-11.8          -> upstream/release-11.8
 * [new branch]            ringtone-asaad        -> upstream/ringtone-asaad
 * [new branch]            test-only             -> upstream/test-only
 * [new branch]            thumbnail-preview     -> upstream/thumbnail-preview
Fetching upstream
hostfile_replace_entries: mkstemp: Read-only file system
update_known_hosts: hostfile_replace_entries failed for /app/.ssh/known_hosts: Read-only file system
+++ Updating remotes done...
+++ Creating local branch automated-cherry-pick-of-mattermost-#34678-upstream-release-11.7-1779087527
Switched to a new branch 'automated-cherry-pick-of-mattermost-#34678-upstream-release-11.7-1779087527'
Branch 'automated-cherry-pick-of-mattermost-#34678-upstream-release-11.7-1779087527' set up to track remote branch 'release-11.7' from 'upstream'.

+++ About to attempt cherry pick of PR #34678 with merge commit 65500cc820a9b3d1c3f86e351598bde99a28ba5f.

Auto-merging server/channels/app/platform/service.go
Auto-merging server/public/model/feature_flags.go
CONFLICT (content): Merge conflict in server/public/model/feature_flags.go
error: could not apply 65500cc820... MM Build 2025 - Skadoosh - Serve plugin metrics to standard /metrics endpoint (#34678)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

+++ Conflicts detected:

UU server/public/model/feature_flags.go
Aborting.

+++ Aborting in-progress git cherry-pick.

+++ Returning you to the master branch and cleaning up.

```

mattermost-build commented:
Looks like you don't have permissions to trigger this command.
 Only available for the PR submitter and org members

amyblais commented:
@lieut-data You previously suggested backporting this to v11.7 for a customer once this has been internally tested. When the testing is done, can you help open a manual cherry-pick PR to ``release-11.7`` branch?

lieut-data commented:
Thanks, @amyblais -- this is still in my backlog to validate on Community, will cherry-pick once confirmed.

amyblais commented:
> Thanks,  -- this is still in my backlog to validate on Community, will cherry-pick once confirmed.

Thanks!
